### ISSUE TITLE
Better diagnostics for `RUF039`

### ISSUE THREAD

**Issue by InSyncWithFoo** [astral-sh/ruff#16713]
This snippet has 10 [`RUF039`](https://docs.astral.sh/ruff/rules/unraw-re-pattern/) diagnostics ([playground](https://play.ruff.rs/81703fb0-ac6f-4b0e-8086-82ed8feabf14)), of which only two can be automatically fixed:

```py
re.compile(
    "["
    "\U0001F600-\U0001F64F"  # emoticons
    "\U0001F300-\U0001F5FF"  # symbols & pictographs
    "\U0001F680-\U0001F6FF"  # transport & map symbols
    "\U0001F1E0-\U0001F1FF"  # flags (iOS)
    "\U00002702-\U000027B0"
    "\U000024C2-\U0001F251"
    "\u200d"  # zero width joiner
    "\u200c"  # zero width non-joiner
    "]+",
    flags=re.UNICODE,
)
```

There are a few improvements that can be made:

* The current logic [does not provide a fix](https://github.com/astral-sh/ruff/blob/acf35c55f8b8f5ce16ac388e7f6a1af884edbe16/crates/ruff_linter/src/rules/ruff/rules/unraw_re_pattern.rs#L173) for string literals that contain backslashes.

  However, save for `\b`, which can either mean "word boundary" or "backspace" depending on context, and `\N{}`, which is not supported by `re` until 3.8, all other escape sequences [are supported](https://docs.python.org/3/library/re.html#regular-expression-syntax) and mean the same things as they would in normal strings. Therefore, while the fix will change the actual runtime representation (`re.compile(r'\u00A0') != re.compile('\u00A0')`), the regular expression semantics will be retained. In such cases, Ruff should offer an unsafe fix.

* The number of diagnostics is not ideal.

  When a string is implicitly concatenated, only one diagnostic should be emitted for and encompassing all parts; the fix should too fix all of them at once. Ruff should only resort to multiple diagnostics in cases where only some of the parts are not raw.

(This issue is a follow-up to #16644.)

**Comment by CrowbarKZ** [comment]
Hi @InSyncWithFoo . I have been looking into this issue and, after experimenting locally a bit, I think I disagree with one part of the requirements above, this one:

> Ruff should only resort to multiple diagnostics in cases where only some of the parts are not raw.

It feels strange to have multiple diagnostics and suggested fixes for 1 concatenated string, e.g. turning something like
```
re.compile(
    r"["
    "\U0001F1E0-\U0001F1FF"
    "\b0001F1E0-\U0001F1FF"
    "]+",
    flags=re.UNICODE,
)
```

Into  (if we apply only safe fixes)
```
re.compile(
    r"["
    "\U0001F1E0-\U0001F1FF"
    "\b0001F1E0-\U0001F1FF"
    r"]+",
    flags=re.UNICODE,
)
```

Does not seem to make the code more readable. Wouldn't it be better to always treat the whole concatenated string as 1 issue, and only apply safe/unsafe fixes when they are applicable to all parts?

**Comment by InSyncWithFoo** [comment]
@CrowbarKZ No strong feelings on my part. Feel free to make decisions as you see fit.

**Comment by robsdedude** [comment]
I'm currently trying to cook up a PR for this.

**Comment by ntBre** [comment]
We closed this in https://github.com/astral-sh/ruff/pull/19065, which added autofixes in more cases but didn't change the number of diagnostics for concatenated literals. I think the number of diagnostics is okay, but we can reopen this if anyone feels strongly about that part.
